lib/daru/index.rb

```diff
@@ -419,4 +501,109 @@ def nils_counted array
       }
     end
   end
-end
+
+  class CategoricalIndex < Index
+    
+    include Enumerable
+    
+    attr_reader :cat_hash, :array, :map_cat_int, :map_int_cat
+    
+    def initialize indexes
+      # Create a hash to map each category to positional indexes
+      @cat_hash = Hash.new []
+      # To link every instance to its category,
+      # it stores integer for every instance representing its category
+      @array = []
+      # Map each category to a unique integer for effective storage in @array
+      map_cat_int = {}
+      # Inverse mapping of @map
+      @map_int_cat = {}
+      
+      cat_count = 0
+      indexes.each_with_index do |index, pos|
+        unless map_cat_int.include? index
+          map_cat_int[index] = cat_count
+          @map_int_cat[cat_count] = index
+          cat_count += 1
+        end
+        @cat_hash[index] += [pos]
+        @array << map_cat_int[index]
+      end
```

Comment:
If I understand correctly what this code should do, it can be rewritten this way:

```ruby
categories = indexes.each_with_index.group_by(&:first)
# %i[a b c a a i b c].each_with_index.group_by(&:first)
# => {:a=>[[:a, 0], [:a, 3], [:a, 4]], :b=>[[:b, 1], [:b, 6]], :c=>[[:c, 2], [:c, 7]], :i=>[[:i, 5]]} 

@cat_hash = categories.map { |cat, group| [cat, group.map(&:last)] }.to_h
# => {:a=>[0, 3, 4], :b=>[1, 6], :c=>[2, 7], :i=>[5]}
map_cat_int = categories.keys.each_with_index.to_h
#  => {:a=>0, :b=>1, :c=>2, :i=>3} 
@map_int_cat = map_cat_int.invert
#  => {0=>:a, 1=>:b, 2=>:c, 3=>:i} 
@array = map_cat_int.values_at(*indexes)
# => [0, 1, 2, 0, 0, 3, 1, 2]
```

The code above is more ruby-ish in the sense it demonstrates _intention_ and _method_ of each variable creation.